core/chains/evm/txmgr/orm.go

```diff
@@ -943,3 +957,259 @@ func (o *orm) SaveReplacementInProgressAttempt(oldAttempt EthTxAttempt, replacem
+		return errors.Wrap(err, "UpdateEthTxUnstartedToInProgress failed to update eth_tx")
+	})
+}
+
+// GetEthTxInProgress returns either 0 or 1 transaction that was left in
+// an unfinished state because something went screwy the last time. Most likely
+// the node crashed in the middle of the ProcessUnstartedEthTxs loop.
+// It may or may not have been broadcast to an eth node.
+func (o *orm) GetEthTxInProgress(fromAddress common.Address, qopts ...pg.QOpt) (etx *EthTx, err error) {
+	qq := o.q.WithOpts()
+	etx = new(EthTx)
+	err = qq.Transaction(func(tx pg.Queryer) error {
+		err = qq.Get(etx, `SELECT * FROM eth_txes WHERE from_address = $1 and state = 'in_progress'`, fromAddress.Bytes())
+		if errors.Is(err, sql.ErrNoRows) {
+			return err
+		} else if err != nil {
+			return errors.Wrap(err, "GetEthTxInProgress failed while loading eth tx")
+		}
+		if err = o.LoadEthTxAttempts(etx, pg.WithQueryer(tx)); err != nil {
+			return errors.Wrap(err, "GetEthTxInProgress failed while loading EthTxAttempts")
+		}
+		if len(etx.EthTxAttempts) != 1 || etx.EthTxAttempts[0].State != EthTxAttemptInProgress {
+			return errors.Errorf("invariant violation: expected in_progress transaction %v to have exactly one unsent attempt. "+
+				"Your database is in an inconsistent state and this node will not function correctly until the problem is resolved", etx.ID)
+		}
+		return nil
+	})
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, nil
+	}
+
+	return etx, errors.Wrap(err, "getInProgressEthTx failed")
+}
+
+func (o *orm) HasInProgressTransaction(account common.Address, chainID big.Int, qopts ...pg.QOpt) (exists bool, err error) {
+	qq := o.q.WithOpts(qopts...)
+	err = qq.Get(&exists, `SELECT EXISTS(SELECT 1 FROM eth_txes WHERE state = 'in_progress' AND from_address = $1 AND evm_chain_id = $2)`, account, chainID.String())
+	return exists, errors.Wrap(err, "hasInProgressTransaction failed")
+}
+
+func (o *orm) UpdateEthKeyNextNonce(newNextNonce, currentNextNonce uint64, address common.Address, chainID big.Int, qopts ...pg.QOpt) error {
+	qq := o.q.WithOpts(qopts...)
+	return qq.Transaction(func(tx pg.Queryer) error {
+		//  We filter by next_nonce here as an optimistic lock to make sure it
+		//  didn't get changed out from under us. Shouldn't happen but can't hurt.
+		res, err := tx.Exec(`UPDATE evm_key_states SET next_nonce = $1, updated_at = $2 WHERE address = $3 AND next_nonce = $4 AND evm_chain_id = $5`, newNextNonce, time.Now(), address, currentNextNonce, chainID.String())
+		if err != nil {
+			return errors.Wrap(err, "NonceSyncer#fastForwardNonceIfNecessary failed to update keys.next_nonce")
+		}
+		rowsAffected, err := res.RowsAffected()
+		if err != nil {
+			return errors.Wrap(err, "NonceSyncer#fastForwardNonceIfNecessary failed to get RowsAffected")
+		}
+		if rowsAffected == 0 {
+			return ErrZeroRowsAffected
```

Comment:
```suggestion
			return errors.Wrap(sql.ErrNoRows, ErrZeroRowsAffected)
```

Gives you broader compatibility.

Reply:
You may want a more specific error here.

Reply:
From the docs it says  `ErrNoRows is returned by Scan when QueryRow doesn't return a row` so we shouldn't combine such an error with zero rows being *affected*

Reply:
Makes sense.

> You may want a more specific error here.

^^ Still recommend this since it solves the inconsistency